clients/pkg/logentry/stages/eventlogmessage.go

```diff
@@ -0,0 +1,164 @@
+func parseEventLogMessageConfig(config interface{}) (*EventLogMessageConfig, error) {
+	cfg := &EventLogMessageConfig{}
+	err := mapstructure.Decode(config, cfg)
+	if err != nil {
+		return nil, err
+	}
+	return cfg, nil
+}
+
+// Ensure a event log message configuration object is valid, checking that any specified source
+// is a valid label name, and setting default values if a nil config object is provided
+func validateEventLogMessageConfig(c *EventLogMessageConfig) error {
+	if c == nil {
+		return errors.New(ErrEmptyEvtLogMsgStageConfig)
+	}
+	if c.Source != nil && !model.LabelName(*c.Source).IsValid() {
+		return fmt.Errorf(ErrInvalidLabelName, *c.Source)
+	}
+	return nil
+}
+
+func (m *eventLogMessageStage) Run(in chan Entry) chan Entry {
+	out := make(chan Entry)
+	key := defaultSource
+	if m.cfg.Source != nil {
+		key = *m.cfg.Source
+	}
+	go func() {
+		defer close(out)
+		for e := range in {
+			err := m.processEntry(e.Extracted, key)
+			if err != nil {
+				continue
+			}
+			out <- e
+		}
+	}()
+	return out
+}
+
+// Process a event log message from extracted with the specified key, adding additional
+// entries into the extracted map
+func (m *eventLogMessageStage) processEntry(extracted map[string]interface{}, key string) error {
+	value, ok := extracted[key]
+	if !ok {
+		if Debug {
+			level.Debug(m.logger).Log("msg", "source not in the extracted values", "source", key)
+		}
+		return nil
+	}
+	s, err := getString(value)
+	if err != nil {
+		if Debug {
+			level.Debug(m.logger).Log("msg", "invalid label value parsed", "value", value)
+		}
+		return err
+	}
+	lines := strings.Split(s, "\r\n")
+	for _, line := range lines {
+		parts := regexp.MustCompile(": ?").Split(line, 2)
+		if len(parts) < 2 {
+			if Debug {
+				level.Debug(m.logger).Log("msg", "invalid line parsed from message", "line", line)
+			}
+			continue
+		}
+		mkey := parts[0]
+		if !model.LabelName(mkey).IsValid() {
+			if m.cfg.DropInvalidLabels {
+				if Debug {
+					level.Debug(m.logger).Log("msg", "invalid label parsed from message", "key", mkey)
+				}
+				continue
+			}
+			mkey = SanitizeLabelName(mkey)
+		}
+		if _, ok := extracted[mkey]; ok && !m.cfg.OverwriteExisting {
+			level.Info(m.logger).Log("msg", "extracted label that already existed, ignoring value from source",
+				"key", mkey, "oldval", extracted[mkey], "newval", parts[1])
+			continue
+		}
+		mval := parts[1]
+		if !model.LabelValue(mval).IsValid() {
+			if Debug {
+				level.Debug(m.logger).Log("msg", "invalid value parsed from message", "value", mval)
+			}
+			continue
+		}
+		extracted[mkey] = mval
+	}
+	if Debug {
+		level.Debug(m.logger).Log("msg", "extracted data debug in event_log_message stage",
+			"extracted data", fmt.Sprintf("%v", extracted))
+	}
+	return nil
+}
+
+func (m *eventLogMessageStage) Name() string {
+	return StageTypeEventLogMessage
+}
+
+// Sanitize a input string to convert it into a valid prometheus label
+func SanitizeLabelName(input string) string {
```

Comment:
We should have this method somewhere already 😊

Reply:
I found [this function](https://github.com/prometheus/prometheus/blob/c70d85baed260f6013afd18d6cd0ffcac4339861/util/strutil/strconv.go#L41) in Prometheus - and I saw it is being used elsewhere in promtail - but it has a (admittedly subtle) bug: it allows `0key` as a label, yet the first character of a Prometheus label must be `[a-zA-Z_]`. Happy to push this fix upstream (though I can't see a way of doing it with regex replace unless we use two regexes) - but I'd prefer not to depend on it for the time being.

Reply:
I've [submitted a PR](https://github.com/prometheus/prometheus/pull/11936) for my fix upstream - once it is accepted, I'll remove this code and import it.